config.md

```diff
@@ -25,19 +25,61 @@ Changing it means creating a new derived image, instead of changing the existing
-This is called a `ChainID`.
-For a single layer (or the layer at the bottom of a stack), the
-`ChainID` is equal to the layer's `DiffID`.
-Otherwise the `ChainID` is given by the formula:
-`ChainID(layerN) = SHA256hex(ChainID(layerN-1) + " " + DiffID(layerN))`.
+While a layer's `DiffID` identifies a single changeset, the `ChainID` identifies the subsequent application of those changesets.
+This ensures that we have handles referring to both the layer itself, as well as the result of the application of a series of changesets.
+Use in combination with `rootfs.diff_ids` while applying layers to a root filesystem to uniquely and safely identify the result.
+
+#### Definition
+
+The `ChainID` of an applied set of layers is defined with the following recursion:
+
+```
+ChainID(L₀) =  DiffID(L₀)
+ChainID(L₀|...|Lₙ₋₁|Lₙ) = Digest(ChainID(L₀|...|Lₙ₋₁) + " " + DiffID(Lₙ))
+```
+
+For this, we define the binary `|` operation to be the result of applying the right operand to the left operand.
+For example, given base layer `A` and a changeset `B`, we refer to the result of applying `B` to `A` as `A|B`.
+
+Above, we define the `ChainID` for a single layer (`L₀`) as equivalent to the `DiffID` for that layer.
+Otherwise, the `ChainID` for `L₀|...|Lₙ₋₁|Lₙ` is defined as recursion `Digest(ChainID(L₀|...|Lₙ₋₁) + " " + DiffID(Lₙ))`.
+
+#### Explanation
+
+Let's say we have layers A, B, C, ordered from bottom to top, where A is the base and C is the top.
+Defining `|` as a binary application operator, the root filesystem may be `A|B|C`.
+While it is implied that `C` is only useful when applied to `A|B`, the identifier `C` is insufficient to identify this result, as we'd have the equality `C = A|B|C`, which isn't true.
+
+The main issue is when we have two definitions of `C`, `C = C` and `C = A|B|C`. If this is true (with some handwaving), `C = x|C` where `x = any application` must be true.
+This means that if an attacker can define `x`, relying on `C` provides no guarantee that the layers were applied in any order.
+
+The `ChainID` addresses this problem by being defined as a compound hash.
+__We differentiate the changeset `C`, from the order dependent application `A|B|C` by saying that the resulting rootfs is identified by ChainID(A|B|C), which can be calculated by `ImageConfig.rootfs`.__
+
+Let's expand the definition of `ChainID(A|B|C)` to explore its internal structure:
+
+```
+ChainID(A) = DiffID(A)
+ChainID(A|B) = Digest(ChainID(A) + " " + DiffID(B))
+ChainID(A|B|C) = Digest(ChainID(A|B) + " " + DiffID(C))
+```
+
+We can replace the each definition and reduce to a single equality:
+
+```
+ChainID(A|B|C) = Digest(Digest(DiffID(A) + " " + DiffID(B)) + " " + DiffID(C))
+```
+
+Hopefully, the above is illustrative of the _actual_ contents of the `ChainID`.
+Most importantly, we can easily see that `ChainID(C) != ChainID(A|B|C)`, otherwise, `ChainID(C) = DiffID(C)`, which is the base case, could not be true.
```

Comment:
> > it's not clear to me why anyone would attempt to identify that array with ChainID(C).
>
> In fact, ChainID(C) = ChainID(∅|C)…

To have `∅|C` you'd need an empty-tar layer.  And `ChainID(C) = DiffID(C) ≠ Digest(DiffID(∅) + " " + DiffID(C)) = ChainID(∅|C)`.  There may be some benefit to a `ChainID` definition that is [based on the assembled filesystem][1] (you get the same `ChainID` for `A|B|C` and `D|E` if the resulting filesystems happen to be identical), but any `ChainID` that is defined recursively on the constituent layers is not actually based on the resulting filesystem.

> It is the chain id for identifying the application of C to the empty file system.

This is true, but that's not what `ChainID(∅|C)` is.  And either way, I [still][2] don't see why anyone would be using this to identify the layer array `[A, B, C]`.

> If these aspects aren't clear to you, why are commenting with such authority?

Because I think these aspects are clear to me?

> I don't mind further clarifying this, but it makes it hard to have a conversation when you pose your attempt to understand as contrarian statements.

I think pushing back on “this doesn't make sense to me” is a good way to have a discussion that gets us to agreement.  There are alternatives though.  For example, I could file a PR with my own take on clarifying `ChainID` (although it would be [very similar to my December comment][3]).  If that seems like it would be more productive, let me know.  But unless we find a way focus on the places where we aren't currently seeing eye to eye, I don't know how we'll resolve those differences.

[1]: https://github.com/opencontainers/image-spec/pull/586/commits/0fa3cc51c853e161ba6e1af7e92cffd77dac9871#diff-c9c91c29b41257aea3a3403cc606ad99R50
[2]: https://github.com/opencontainers/image-spec/pull/586#discussion_r103807307
[3]: https://github.com/opencontainers/image-spec/issues/482#issuecomment-265565694